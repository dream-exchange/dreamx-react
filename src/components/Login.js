import React, { Component } from "react";
import { connect } from "react-redux";
import Loading from "./Loading";

import { accountLogin, accountLogout } from "../actions";
import "./Login.scss";
import Button from "./Button";

const metamask = require("../images/metamask-icon.svg");

class Login extends Component {
  render() {
    if (this.props.account.loading) {
      return (
        <div className={`Login card text-center ${this.props.app.theme}`}>
          <Loading
            active={this.props.account.loading}
            type="absolute"
            theme={this.props.app.theme}
          />
        </div>
      );
    }

    if (this.props.account.address) {
      const address_first_four_digits = this.props.account.address.substring(
        2,
        6
      );
      const address_last_four_digits = this.props.account.address.substring(
        this.props.account.address.length - 4,
        this.props.account.address.length
      );

      return (
        <div className={`Login card text-center ${this.props.app.theme}`}>
          <div className="card-body">
            <h5 className="card-title">Welcome!</h5>
            <p className="card-text">
              You have now logged in. Your address is 0x
              {address_first_four_digits}...{address_last_four_digits}
            </p>
            <Button
              theme={this.props.app.theme}
              onClick={this.props.accountLogout}
            >
              Disconnect
            </Button>
          </div>
        </div>
      );
    }

    if (this.props.account.metamask === "unavailable") {
      return (
        <div className={`Login card text-center ${this.props.app.theme}`}>
          <div className="card-body">
            <h5 className="card-title">MetaMask is not installed</h5>
            <p className="card-text">Would you like to install it now?</p>
            <a
              href="https://metamask.io"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Button theme={this.props.app.theme}>Install MetaMask</Button>
            </a>
          </div>
        </div>
      );
    }

    if (this.props.account.metamask === "wrongnetwork") {
      return (
        <div className={`Login card text-center ${this.props.app.theme}`}>
          <div className="card-body">
            <h5 className="card-title">Wrong network detected</h5>
            <p className="card-text">
              Please set MetaMask's network to {this.props.exchange.networkName}{" "}
              and try again.
            </p>
            <Button
              onClick={this.props.accountLogin}
              theme={this.props.app.theme}
            >
              Try again
            </Button>
          </div>
        </div>
      );
    }

    return (
      <div className={`Login card text-center ${this.props.app.theme}`}>
        <div className="card-body">
          <h5 className="card-title">You are not logged in</h5>
          <p className="card-text">
            Connect to MetaMask to manage your account.
          </p>
          <Button
            onClick={this.props.accountLogin}
            theme={this.props.app.theme}
          >
            Connect
          </Button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ account, app }) => {
  return { account, app };
};

const mapActionsToProps = {
  accountLogin,
  accountLogout
};

export default connect(
  mapStateToProps,
  mapActionsToProps
)(Login);
