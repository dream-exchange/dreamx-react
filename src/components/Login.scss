$shadow: 0px 5px 15px rgba(1, 1, 1, 0.1);
$light-text: rgba(0, 0, 0, 0.54);
$light-primary: #303f9f;
$light-secondary: #f5f5f5;
$light-secondary-tint: white;
$light-secondary-shade: #dcdcdc;
$dark-text: hsla(0,0%,100%,.7);
$dark-primary: #EA6C2C;
$dark-secondary: #232d44;
$dark-secondary-tint: #384256;
$dark-secondary-shade: #1f283d;

.Login {
  margin-top: 30px;
  box-shadow: $shadow;
  border: 0px;
  box-sizing: border-box;
  min-height: 200px;
  &.light {
    background-color: $light-secondary-tint;
    color: $light-text;
  }
  &.dark {
    background-color: $dark-secondary-tint;
    color: $dark-text;
  }
  .card-body{
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
  }
}