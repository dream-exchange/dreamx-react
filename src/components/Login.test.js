import React from "react";
import { Provider } from "react-redux";
import {
  render,
  fireEvent,
  cleanup,
  waitForElement
} from "react-testing-library";

import Login from "./Login";
import store from "../store";
import {
  APP_INITIALIZE,
  ACCOUNT_LOGIN,
  RESET_INITIAL_STATE
} from "../actions/types";

afterEach(() => {
  store.dispatch({
    type: RESET_INITIAL_STATE
  });
  global.ethereum = undefined;
  cleanup();
});

it("renders without crashing", () => {
  render(
    <Provider store={store}>
      <Login />
    </Provider>
  );
});

it("when not logged in", () => {
  const { container } = render(
    <Provider store={store}>
      <Login />
    </Provider>
  );

  expect(container).toMatchSnapshot();
});

it("when logged in", async () => {
  global.ethereum = {
    networkVersion: "3",
    isMetaMask: true,
    enable: jest.fn(() => {
      return new Promise((resolve, reject) => {
        resolve(["0x4e30dba9762aba125f5ab81647edebff9f9df7a7"]);
      });
    })
  };
  store.dispatch({
    type: APP_INITIALIZE,
    payload: {
      address: "0x45717b73011AD0fB45bF70a336563B32873E7b23",
      networkId: "3",
      networkName: "Ropsten"
    }
  });

  const { getByText, container } = render(
    <Provider store={store}>
      <Login />
    </Provider>
  );

  fireEvent.click(getByText(/^Connect$/i));

  expect(await waitForElement(() => getByText(/Welcome!/i))).toBeTruthy();

  expect(container).toMatchSnapshot();
});

it("when metamask is not installed", async () => {
  global.ethereum = undefined;

  const { getByText, container } = render(
    <Provider store={store}>
      <Login />
    </Provider>
  );

  fireEvent.click(getByText(/^Connect$/i));

  expect(
    await waitForElement(() => getByText(/MetaMask is not installed/i))
  ).toBeTruthy();

  expect(container).toMatchSnapshot();
});

it("when metamask is on the wrong network", async () => {
  global.ethereum = {
    networkVersion: "1",
    isMetaMask: true,
    enable: jest.fn(() => {
      return new Promise((resolve, reject) => {
        resolve(["0x4e30dba9762aba125f5ab81647edebff9f9df7a7"]);
      });
    })
  };
  store.dispatch({
    type: APP_INITIALIZE,
    payload: {
      address: "0x45717b73011AD0fB45bF70a336563B32873E7b23",
      networkId: "3",
      networkName: "Ropsten"
    }
  });

  const { getByText, container } = render(
    <Provider store={store}>
      <Login />
    </Provider>
  );

  fireEvent.click(getByText(/^Connect$/i));

  expect(
    await waitForElement(() => getByText(/Wrong network detected/i))
  ).toBeTruthy();

  expect(container).toMatchSnapshot();
});

// it("when log out", async () => {
//   store.dispatch({
//     type: ACCOUNT_LOGIN,
//     payload: { address: "0x4E30dBA9762ABa125f5Ab81647eDEbFF9f9DF7a7" }
//   });

//   const { getByText, container } = render(
//     <Provider store={store}>
//       <Login />
//     </Provider>
//   );

//   fireEvent.click(getByText(/Disconnect/i));

//   expect(
//     await waitForElement(() =>
//       getByText(/Connect to MetaMask to manage your account/i)
//     )
//   ).toBeTruthy();

//   expect(container).toMatchSnapshot();
// });
