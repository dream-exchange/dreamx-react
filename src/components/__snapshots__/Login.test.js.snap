// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when log out 1`] = `
<div>
  <div
    class="Login card text-center dark"
  >
    <div
      class="card-body"
    >
      <h5
        class="card-title"
      >
        You are not logged in
      </h5>
      <p
        class="card-text"
      >
        Connect to MetaMask to manage your account.
      </p>
      <div
        class="Button dark"
      >
        Connect
      </div>
    </div>
  </div>
</div>
`;

exports[`when logged in 1`] = `
<div>
  <div
    class="Login card text-center dark"
  >
    <div
      class="card-body"
    >
      <h5
        class="card-title"
      >
        Welcome!
      </h5>
      <p
        class="card-text"
      >
        You have now logged in. Your address is 0x
        4e30
        ...
        f7a7
      </p>
      <div
        class="Button dark"
      >
        Disconnect
      </div>
    </div>
  </div>
</div>
`;

exports[`when metamask is not installed 1`] = `
<div>
  <div
    class="Login card text-center dark"
  >
    <div
      class="card-body"
    >
      <h5
        class="card-title"
      >
        MetaMask is not installed
      </h5>
      <p
        class="card-text"
      >
        Would you like to install it now?
      </p>
      <a
        href="https://metamask.io"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="Button dark"
        >
          Install MetaMask
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`when metamask is on the wrong network 1`] = `
<div>
  <div
    class="Login card text-center dark"
  >
    <div
      class="card-body"
    >
      <h5
        class="card-title"
      >
        Wrong network detected
      </h5>
      <p
        class="card-text"
      >
        Please set MetaMask's network to 
        Ropsten
         and try again.
      </p>
      <div
        class="Button dark"
      >
        Try again
      </div>
    </div>
  </div>
</div>
`;

exports[`when not logged in 1`] = `
<div>
  <div
    class="Login card text-center dark"
  >
    <div
      class="card-body"
    >
      <h5
        class="card-title"
      >
        You are not logged in
      </h5>
      <p
        class="card-text"
      >
        Connect to MetaMask to manage your account.
      </p>
      <div
        class="Button dark"
      >
        Connect
      </div>
    </div>
  </div>
</div>
`;
